torch/csrc/jit/export.cpp

```diff
@@ -220,19 +220,123 @@ void encodeModel(onnx::ModelProto* p_m, const std::shared_ptr<Graph>& g,
+
+    // TODO: switch ATen tracing to not insert selects for single output.
+    auto* rhs = n->inputs().at(n->inputs().size() - 1);
+
+    // The rhs input isn't actually an expand, so no fusion available
+    if (rhs->kind() != kExpand) continue;
+
+    auto* new_rhs = rhs->input();
+
+    // We need to know what the type pre-expand is.  We should basically
+    // always have this information (because expands are only ever traced,
+    // not generated from symbolic), but if for some reason we don't
+    // have it, we need to skip.
+    if (!new_rhs->hasType()) continue;
+
+    // Not all broadcasts are supported by ONNX broadcast.
+    if (!fusibleExpandTo(new_rhs->type()->expect<TensorType>()->sizes(),    // from
+                         rhs->type()->expect<TensorType>()->sizes()) // to
+       ) continue;
+
+    auto *new_n = graph->createClone(n, [&](Node* n) { return n == rhs ? new_rhs : n; });
+    new_n->i_(kbroadcast, 1);
+    new_n->insertAfter(n);
+    n->replaceAllUsesWith(new_n);
+    it.destroyCurrent();
+    if (rhs->uses().size() == 0) {
+      if (*it == rhs) {
+        it.destroyCurrent();
+      } else {
+        rhs->destroy();
+      }
+    }
+  }
+}
+
+
 void standardizeGraph(const std::shared_ptr<Graph>& graph) {
+  // TODO: move this out of here...
+  fuseBroadcast(graph);
+
   for (auto it = graph->nodes().begin(); it != graph->nodes().end(); ++it) {
+      // Macro'ed so we get a marginally better line number on failed export
 #define FAIL_EXPORT(name) \
-      throw std::runtime_error(std::string("Couldn't export ") + name + " function - " \
-              "maybe it doesn't implement a symbolic definition?");
+      throw std::runtime_error(std::string("ONNX export failed: ") + name + "\n\nGraph we tried to export:\n" + graph->toString());
     IR_IF(*it, CppOp)
       auto cpp_node = static_cast<torch::jit::CppOp*>(value);
-      FAIL_EXPORT(cpp_node->name())
+      FAIL_EXPORT("Couldn't export C++ operator " + cpp_node->name())
     IR_ELSEIF(PythonOp)
       auto py_node = static_cast<torch::jit::PythonOp*>(value);
-      FAIL_EXPORT(py_node->name())
+      FAIL_EXPORT("Couldn't export Python operator " + py_node->name())
     IR_ELSE()
-      // Do nothing.
+      // Expand is not a real ONNX operator yet, reject it
+      if (it->kind() == kExpand) {
+        FAIL_EXPORT("Couldn't export operator expand; this usually means you used a form of broadcasting that ONNX does not currently support");
+      }
+      if (it->kind() == kUndefined) {
+        FAIL_EXPORT("Couldn't export undefined constant tensor (please file an issue)")
+      }
+      std::string n = symbolToString(it->kind());
+      if (n.size() == 0) {
+        FAIL_EXPORT("Operator to export had empty name (please file an issue)")
+      }
+      // NB: Upper-case is ONNX, lower-case is ATen.  If we want to be more
+      // robust, need to explicitly flag operators as ONNX or ATen
+      if (!isupper(n[0])) {
```

Comment:
Great, we can catch the missing op, such as cat

Reply:
Hm, can we do a whitelist of Aten operators actually? It should be easily autogenerated. Better yet - both ONNX and ATen. Implicit conversion is pretty confusing to users.

Reply:
I would much rather do the proposal here: https://github.com/ezyang/pytorch/issues/119